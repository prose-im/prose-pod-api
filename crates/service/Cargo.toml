[package]
name = "service"
version = { workspace = true }
edition = { workspace = true }
publish = false
license = { workspace = true }
authors = ["RÃ©mi Bardon <remi@remibardon.name>"]

[dependencies]
anyhow = { workspace = true }
async-trait = { workspace = true }
base64 = { workspace = true }
chrono = { workspace = true }
email_address = { workspace = true }
figment = { workspace = true }
form_urlencoded = { workspace = true }
futures = { workspace = true }
linked_hash_set = { workspace = true }
hickory-proto = { workspace = true }
hickory-resolver = { workspace = true }
hmac = { workspace = true }
iso8601-duration = { workspace = true }
jid = { workspace = true }
lazy_static = { workspace = true }
lettre = { workspace = true }
mime = { workspace = true }
minidom = { workspace = true }
parking_lot = { workspace = true }
prose-xmpp = { workspace = true }
prosody-config = { workspace = true }
rand = { workspace = true }
reqwest = { workspace = true }
sea-orm = { workspace = true }
secrecy = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
serde_with = { workspace = true }
sha2 = { workspace = true }
strum = { workspace = true }
tempfile = { workspace = true }
thiserror = { workspace = true }
tokio = { workspace = true }
tokio-util = { workspace = true }
tracing = { workspace = true }
url_serde = { workspace = true }
urlencoding = { workspace = true }
uuid = { workspace = true, features = ["v4"] }
xmpp-parsers = { workspace = true }

[dev-dependencies]
cucumber = { workspace = true }
insta = { workspace = true }
migration = { workspace = true }

[[test]]
name = "behavior"
harness = false
