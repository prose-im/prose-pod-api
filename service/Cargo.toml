[package]
name = "service"
version = "0.1.0"
edition = "2021"
publish = false
license = "MPL-2.0"
authors = ["RÃ©mi Bardon <remi@remibardon.name>"]

[dependencies]
chrono = "0.4"
entity = { path = "../entity" }
figment = { version = "0.10", features = ["toml"] }
lettre = { version = "0.11", features = [
    "smtp-transport",
    "rustls-tls",
    "hostname",
    "builder",
], default-features = false }
log = { version = "0.4", features = ["std"] }
prosody-config = { path = "../prosody-config" }
reqwest = { version = "0.12", features = [
    "blocking",
    "rustls-tls",
    "json",
    "trust-dns",
], default-features = false }
sea-orm = { version = "1.0.0-rc.3", features = [
    "runtime-tokio-rustls",
    "sqlx-sqlite",
    #"macros",
] }
serde = { version = "1", features = ["derive"] }
tempfile = "3"
tokio = "1"
urlencoding = "2"
url_serde = "0.2"
uuid = { version = "1", features = ["v4"] }
vcard_parser = "0.2"

[dev-dependencies]
cucumber = "0.21"
insta = "1"
migration = { path = "../migration" }

[[test]]
name = "behavior"
harness = false
