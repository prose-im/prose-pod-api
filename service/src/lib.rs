// prose-pod-api
//
// Copyright: 2023–2024, Rémi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

pub extern crate prose_xmpp;
pub extern crate xmpp_parsers;

mod auth_service;
pub mod config;
pub mod dependencies;
mod jwt_service;
mod mutation_error;
pub mod prosody;
pub mod repositories;
pub mod server_ctl;
pub mod util;
pub mod xmpp;

pub use auth_service::{
    AuthError, AuthService, AuthServiceImpl, LiveAuthService, JWT_PROSODY_TOKEN_KEY,
};
use config::Config;
pub use dependencies::any_notifier as notifier;
pub use entity::model::{EmailAddress, JIDNode, MemberRole};
use entity::server_config::Model as ServerConfig;
pub use jwt_service::{JWTError, JWTKey, JWTService, JWT_JID_KEY};
pub use mutation_error::*;
pub use prosody::prosody_config_from_db;
pub use reqwest::Client as HttpClient;
pub use server_ctl::ServerCtl;
pub use xmpp::xmpp_service::{self, *};

pub use prosody_config::ProsodyConfigSection;
pub use sea_orm;

// TODO: Move all code which depends on this into `service`
pub mod deprecated {
    pub use entity::{
        model::{DateLike, Duration, PossiblyInfinite},
        server_config::ActiveModel as ServerConfigActiveModel,
        workspace::ActiveModel as WorkspaceActiveModel,
    };
}

trait ProseDefault {
    fn prose_default(server_config: &ServerConfig, app_config: &Config) -> Self;
}
