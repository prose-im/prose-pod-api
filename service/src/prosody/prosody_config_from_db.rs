// prose-pod-api
//
// Copyright: 2024, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use std::ops::Deref;

use entity::model as db;
use entity::server_config::Model as ServerConfig;
use prosody_config::{linked_hash_set::LinkedHashSet, *};
use utils::def;

use crate::config::Config;
use crate::ProseDefault;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct ProsodyConfig(prosody_config::ProsodyConfig);

impl Deref for ProsodyConfig {
    type Target = prosody_config::ProsodyConfig;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl ProsodyConfig {
    pub fn print(&self) -> ProsodyConfigFile {
        self.0.clone().print(vec![
            "Prose Pod Server".into(),
            "XMPP Server Configuration".into(),
            r#"/!\ This file has been automatically generated by Prose Pod API."#.into(),
            r#"/!\ Do NOT edit this file manually or your changes will be overriden during the next reload."#.into(),
        ].into())
    }

    pub fn all_sections(&self) -> Vec<&ProsodySettings> {
        let mut all_sections = vec![&self.global_settings];
        all_sections.extend(
            self.additional_sections
                .iter()
                .map(|section| section.settings()),
        );
        all_sections
    }
    pub fn all_enabled_modules(&self) -> LinkedHashSet<String> {
        self.all_sections()
            .iter()
            .map(|section| section.modules_enabled.clone().unwrap_or_default())
            .reduce(|acc, e| acc.union(&e).cloned().collect())
            .unwrap_or_default()
    }
}

impl ToString for ProsodyConfig {
    fn to_string(&self) -> String {
        self.print().to_string()
    }
}

pub fn prosody_config_from_db(model: ServerConfig, app_config: &Config) -> ProsodyConfig {
    let mut config = prosody_config::ProsodyConfig::prose_default(&model, app_config);

    let global_settings = &mut config.global_settings;
    let muc_settings = config
        .additional_sections
        .iter_mut()
        .find_map(|section| match section {
            ProsodyConfigSection::Component {
                plugin, settings, ..
            } if plugin.as_str() == "muc" => Some(settings),
            _ => None,
        })
        .expect("The 'Chatrooms' section should always be present.");

    if model.message_archive_enabled {
        add_enabled_module(global_settings, "mam");
        global_settings.archive_expires_after =
            Some(model.message_archive_retention.into_prosody());
        global_settings.default_archive_policy = Some(true);
        global_settings.max_archive_query_results = Some(100);
        add_enabled_module(muc_settings, "muc_mam");
    }

    if model.file_upload_allowed {
        config
            .additional_sections
            .push(ProsodyConfigSection::Component {
                hostname: "upload.prose.org.local".into(),
                plugin: "http_file_share".into(),
                name: "HTTP File Upload".into(),
                settings: ProsodySettings {
                    http_file_share_size_limit: Some(Bytes::MebiBytes(20)),
                    http_file_share_daily_quota: Some(Bytes::MebiBytes(250)),
                    http_file_share_expires_after: Some(
                        model.file_storage_retention.into_prosody(),
                    ),
                    http_host: Some("localhost".into()),
                    http_external_url: Some("http://localhost:5280".into()),
                    ..Default::default()
                },
            })
    }

    ProsodyConfig(config)
}

// ===== Default configuration =====

impl ProseDefault for prosody_config::ProsodyConfig {
    fn prose_default(server_config: &ServerConfig, app_config: &Config) -> Self {
        let api_jid = app_config.api_jid();
        let api_jid = JID::try_from(api_jid.to_string()).expect("Invalid JID: {api_jid}");
        Self {
            global_settings: ProsodySettings {
                pidfile: Some("/var/run/prosody/prosody.pid".into()),
                authentication: Some(AuthenticationProvider::InternalHashed),
                storage: Some(StorageConfig::Raw(StorageBackend::Internal)),
                log: Some(LogConfig::Map(
                    vec![
                        (LogLevel::Info, LogLevelValue::Console),
                        (LogLevel::Warn, LogLevelValue::Console),
                        (LogLevel::Error, LogLevelValue::Console),
                    ]
                    .into_iter()
                    .collect(),
                )),
                interfaces: Some(vec![Interface::AllIPv4]),
                c2s_ports: Some(vec![5222]),
                s2s_ports: Some(vec![5269]),
                http_ports: Some(vec![app_config.server.http_port]),
                http_interfaces: Some(vec![Interface::AllIPv4]),
                https_ports: Some(vec![]),
                https_interfaces: Some(vec![]),
                modules_enabled: Some(
                    vec![
                        "auto_activate_hosts",
                        "roster",
                        "groups_internal",
                        "saslauth",
                        "tls",
                        "dialback",
                        "disco",
                        "posix",
                        "smacks",
                        "private",
                        "vcard_legacy",
                        "vcard4",
                        "version",
                        "uptime",
                        "time",
                        "ping",
                        "lastactivity",
                        "pep",
                        "blocklist",
                        "limits",
                        "carbons",
                        "csi",
                        "server_contact_info",
                        "websocket",
                        "s2s_bidi",
                    ]
                    .into_iter()
                    .map(ToString::to_string)
                    .collect(),
                ),
                ssl: Some(SSLConfig::Manual {
                    key: "/etc/prosody/certs/prose.org.local.key".into(),
                    certificate: "/etc/prosody/certs/prose.org.local.crt".into(),
                }),
                allow_registration: Some(false),
                c2s_require_encryption: Some(true),
                s2s_require_encryption: Some(true),
                s2s_secure_auth: Some(false),
                c2s_stanza_size_limit: Some(Bytes::KibiBytes(256)),
                s2s_stanza_size_limit: Some(Bytes::KibiBytes(512)),
                limits: Some(
                    vec![
                        (
                            ConnectionType::ClientToServer,
                            ConnectionLimits {
                                rate: Some(DataRate::KiloBytesPerSec(50)),
                                burst: Some(Duration(TimeLike::Seconds(2))),
                            },
                        ),
                        (
                            ConnectionType::ServerToServerInbounds,
                            ConnectionLimits {
                                rate: Some(DataRate::KiloBytesPerSec(250)),
                                burst: Some(Duration(TimeLike::Seconds(4))),
                            },
                        ),
                    ]
                    .into_iter()
                    .collect(),
                ),
                consider_websocket_secure: Some(true),
                cross_domain_websocket: None,
                contact_info: Some(ContactInfo {
                    admin: vec!["mailto:hostmaster@prose.org.local".to_string()],
                    ..Default::default()
                }),
                upgrade_legacy_vcards: Some(true),
                ..Default::default()
            },
            additional_sections: vec![
                ProsodyConfigSection::VirtualHost {
                    hostname: server_config.domain.to_owned(),
                    settings: ProsodySettings {
                        admins: Some(vec![api_jid.to_owned()].into_iter().collect()),
                        modules_enabled: Some(
                            vec![
                                "rest",
                                "http_oauth2",
                                "admin_rest",
                                // "init_admin",
                            ]
                            .into_iter()
                            .map(ToString::to_string)
                            .collect(),
                        ),
                        http_host: Some(app_config.server.local_hostname.to_owned()),
                        custom_settings: vec![
                            Group::new(
                                "mod_http_oauth2",
                                vec![def(
                                    "allowed_oauth2_grant_types",
                                    vec!["password"],
                                )],
                            ),
                            // // See <https://github.com/prose-im/prose-pod-server/blob/3b54d071880dff669f0193a8068733b089936751/plugins/mod_init_admin.lua>.
                            // Group::new(
                            //     "mod_init_admin",
                            //     vec![def(
                            //         "init_admin_jid",
                            //         api_jid.to_owned(),
                            //     )],
                            // ),
                        ],
                        ..Default::default()
                    },
                },
                ProsodyConfigSection::VirtualHost {
                    hostname: "admin.prose.org.local".into(),
                    settings: ProsodySettings {
                        admins: Some(vec![api_jid.to_owned()].into_iter().collect()),
                        modules_enabled: Some(
                            vec![
                                "admin_rest",
                                "init_admin",
                            ]
                            .into_iter()
                            .map(ToString::to_string)
                            .collect(),
                        ),
                        http_host: Some(app_config.server.local_hostname_admin.to_owned()),
                        custom_settings: vec![
                            // See <https://github.com/prose-im/prose-pod-server/blob/3b54d071880dff669f0193a8068733b089936751/plugins/mod_init_admin.lua>.
                            Group::new(
                                "mod_init_admin",
                                vec![
                                    def("init_admin_jid", api_jid.to_owned()),
                                    def(
                                        "init_admin_password_env_var_name",
                                        "PROSE_API__ADMIN_PASSWORD",
                                    ),
                                ],
                            ),
                        ],
                        ..Default::default()
                    },
                },
                ProsodyConfigSection::Component {
                    hostname: format!("groups.{}", server_config.domain),
                    plugin: "muc".into(),
                    name: "Chatrooms".into(),
                    settings: ProsodySettings {
                        restrict_room_creation: Some(RoomCreationRestriction::DomainOnly),
                        max_archive_query_results: Some(100),
                        muc_log_all_rooms: Some(true),
                        muc_log_expires_after: Some(PossiblyInfinite::Infinite),
                        muc_log_by_default: Some(true),
                        ..Default::default()
                    },
                },
            ],
        }
    }
}

// ===== Convenience methods =====

// TODO: Remove unused code

// fn add_admin(settings: &mut ProsodySettings, jid: JID) {
//     settings
//         .admins
//         .get_or_insert_with(Default::default)
//         .insert(jid);
// }
// fn remove_admin(settings: &mut ProsodySettings, jid: &JID) {
//     settings
//         .admins
//         .get_or_insert_with(Default::default)
//         .remove(jid);
// }

fn add_enabled_module(settings: &mut ProsodySettings, module_name: &'static str) -> bool {
    settings
        .modules_enabled
        .get_or_insert_with(Default::default)
        .insert(module_name.into())
}
// fn remove_enabled_module(settings: &mut ProsodySettings, module_name: &String) -> bool {
//     settings
//         .modules_enabled
//         .get_or_insert_with(Default::default)
//         .remove(module_name)
// }

// fn add_disabled_module(settings: &mut ProsodySettings, module_name: String) -> bool {
//     settings
//         .modules_disabled
//         .get_or_insert_with(Default::default)
//         .insert(module_name)
// }
// fn remove_disabled_module(settings: &mut ProsodySettings, module_name: &String) -> bool {
//     settings
//         .modules_disabled
//         .get_or_insert_with(Default::default)
//         .remove(module_name)
// }

// fn set_rate_limit(settings: &mut ProsodySettings, conn_type: ConnectionType, value: DataRate) {
//     settings
//         .limits
//         .get_or_insert_with(Default::default)
//         .entry(conn_type.into())
//         .or_insert_with(Default::default)
//         .rate = Some(value)
// }
// fn set_burst_limit(
//     settings: &mut ProsodySettings,
//     conn_type: ConnectionType,
//     value: Duration<TimeLike>,
// ) {
//     settings
//         .limits
//         .get_or_insert_with(Default::default)
//         .entry(conn_type.into())
//         .or_insert_with(Default::default)
//         .burst = Some(value)
// }

// ===== Model mappings =====

trait IntoProsody<T> {
    fn into_prosody(self) -> T;
}

impl IntoProsody<DateLike> for db::DateLike {
    fn into_prosody(self) -> DateLike {
        match self {
            Self::Days(n) => DateLike::Days(n),
            Self::Weeks(n) => DateLike::Weeks(n),
            Self::Months(n) => DateLike::Months(n),
            Self::Years(n) => DateLike::Years(n),
        }
    }
}

impl IntoProsody<TimeLike> for db::TimeLike {
    fn into_prosody(self) -> TimeLike {
        match self {
            db::TimeLike::Seconds(_) => todo!(),
            db::TimeLike::Minutes(_) => todo!(),
            db::TimeLike::Hours(_) => todo!(),
        }
    }
}

impl<A, B> IntoProsody<Duration<B>> for db::Duration<A>
where
    A: db::DurationContent + IntoProsody<B>,
    B: DurationContent,
{
    fn into_prosody(self) -> Duration<B> {
        Duration(self.0.into_prosody())
    }
}

impl<A, B> IntoProsody<PossiblyInfinite<B>> for db::PossiblyInfinite<A>
where
    A: IntoProsody<B>,
{
    fn into_prosody(self) -> PossiblyInfinite<B> {
        match self {
            Self::Infinite => PossiblyInfinite::Infinite,
            Self::Finite(d) => PossiblyInfinite::Finite(d.into_prosody()),
        }
    }
}

impl IntoProsody<ConnectionType> for db::ConnectionType {
    fn into_prosody(self) -> ConnectionType {
        match self {
            Self::ClientToServer => ConnectionType::ClientToServer,
            Self::ServerToServerInbounds => ConnectionType::ServerToServerInbounds,
            Self::ServerToServerOutbounds => ConnectionType::ServerToServerOutbounds,
        }
    }
}
