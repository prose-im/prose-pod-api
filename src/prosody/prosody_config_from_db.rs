// prose-pod-api
//
// Copyright: 2024, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use entity::server_config::Model as ServerConfig;

use super::ProsodyConfig;

pub fn prosody_config_from_db(model: ServerConfig) -> ProsodyConfig {
    let mut config = ProsodyConfig::default();

    if model.message_archive_enabled {
        config.modules_enabled.insert("mam".to_string());
        config.archive_expires_after = Some(model.message_archive_retention);
    }

    config
}

// Some old method definitions
// FIXME: Remove useless code below

// fn add_admin(&mut self, jid: JID) {
//     self.config.admins.insert(jid);
// }
// fn remove_admin(&mut self, jid: &JID) {
//     self.config.admins.remove(jid);
// }

// fn add_enabled_module(&mut self, module_name: String) -> bool {
//     self.config.enabled_modules.insert(module_name)
// }
// fn remove_enabled_module(&mut self, module_name: &String) -> bool {
//     self.config.enabled_modules.remove(module_name)
// }

// fn add_disabled_module(&mut self, module_name: String) -> bool {
//     self.config.disabled_modules.insert(module_name)
// }
// fn remove_disabled_module(&mut self, module_name: &String) -> bool {
//     self.config.disabled_modules.remove(module_name)
// }

// fn set_rate_limit(&mut self, conn_type: ConnectionType, value: DataRate) {
//     self.config
//         .limits
//         .entry(conn_type.into())
//         .or_insert_with(Default::default)
//         .rate = Some(value)
// }
// fn set_burst_limit(&mut self, conn_type: ConnectionType, value: DurationTime) {
//     self.config
//         .limits
//         .entry(conn_type.into())
//         .or_insert_with(Default::default)
//         .burst = Some(value)
// }
// fn set_timeout(&mut self, value: DurationTime) {
//     self.config.limits_resolution = Some(value.seconds().clone());
// }

// fn enable_message_archiving(&mut self) -> bool {
//     self.add_enabled_module("mam".to_string())
// }
// fn disable_message_archiving(&mut self) -> bool {
//     self.remove_enabled_module(&"mam".to_string())
// }
