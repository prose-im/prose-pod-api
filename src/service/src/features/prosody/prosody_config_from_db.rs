// prose-pod-api
//
// Copyright: 2024, Rémi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use prosody_config::{utils::def, *};
use tracing::{info, warn};

use crate::{
    app_config::{AppConfig, FILE_SHARE_HOST},
    models::{self as db},
    server_config::ServerConfig,
    ProseDefault,
};

use super::{prosody_config::util::*, prosody_overrides::ProsodyOverrides, ProsodyConfig};

pub fn prosody_config_from_db(model: ServerConfig, app_config: &AppConfig) -> ProsodyConfig {
    let mut config = prosody_config::ProsodyConfig::prose_default(&model, app_config);

    let global_settings = &mut config.global_settings;
    let muc_settings = config
        .additional_sections
        .iter_mut()
        .find_map(|section| match section {
            ProsodyConfigSection::Component {
                plugin, settings, ..
            } if plugin.as_str() == "muc" => Some(settings),
            _ => None,
        })
        .expect("The 'Chatrooms' section should always be present.");

    if model.message_archive_enabled {
        add_enabled_module(global_settings, "mam");
        global_settings.archive_expires_after =
            Some(model.message_archive_retention.into_prosody());
        global_settings.default_archive_policy = Some(ArchivePolicy::Always);
        global_settings.max_archive_query_results = Some(100);
        add_enabled_module(muc_settings, "muc_mam");
    }

    if model.file_upload_allowed {
        config
            .additional_sections
            .push(ProsodyConfigSection::Component {
                hostname: FILE_SHARE_HOST.into(),
                plugin: "http_file_share".into(),
                name: "HTTP File Upload".into(),
                settings: ProsodySettings {
                    http_file_share_size_limit: Some(Bytes::MebiBytes(20)),
                    http_file_share_daily_quota: Some(Bytes::MebiBytes(250)),
                    http_file_share_expires_after: Some(
                        model.file_storage_retention.into_prosody(),
                    ),
                    http_host: Some("localhost".into()),
                    http_external_url: Some("http://localhost:5280".into()),
                    ..Default::default()
                },
            })
    }

    if model.federation_enabled {
        global_settings.unmark_disabled("s2s");
        global_settings.enable_module("s2s_bidi".to_owned());
        global_settings.s2s_ports = Some(vec![5269]);
        global_settings.s2s_require_encryption = Some(true);
        global_settings.s2s_secure_auth = Some(false);
        global_settings.limits.get_or_insert_default().insert(
            ConnectionType::ServerToServerInbounds,
            ConnectionLimits {
                rate: Some(DataRate::KiloBytesPerSec(250)),
                burst: Some(Duration(TimeLike::Seconds(4))),
            },
        );

        if model.federation_whitelist_enabled {
            global_settings.enable_module("s2s_whitelist".to_owned());
            global_settings.s2s_whitelist = Some(model.federation_friendly_servers);
        }
    }

    if model.c2s_unencrypted {
        warn!("Debug config `c2s_unencrypted` is enabled.");
        global_settings.enable_module("reload_modules".to_owned());
        global_settings.custom_settings.push(Group::new(
            "Debug config: c2s_unencrypted",
            vec![
                def("c2s_require_encryption", false),
                def("allow_unencrypted_plain_auth", true),
                def("reload_modules", vec!["saslauth"]),
            ],
        ));
    }

    for module in app_config.prosody.additional_modules_enabled.iter() {
        config.global_settings.enable_module(module.clone());
    }

    if let Some(overrides) = model.prosody_overrides {
        match serde_json::from_value::<ProsodyOverrides>(overrides) {
            Ok(ProsodyOverrides {
                c2s_require_encryption,
            }) => {
                info!("Applying overrides to the generated Prosody configuration file…");
                macro_rules! override_global_setting {
                    ($var:ident) => {
                        if let Some($var) = $var {
                            config.global_settings.$var = Some($var);
                        }
                    };
                }
                override_global_setting!(c2s_require_encryption);
            },
            Err(err) => warn!("Prosody overrides stored in database cannot be read, they won’t be applied. To fix this, call `PUT /v1/server/config/prosody-overrides` with a new value. You can `GET /v1/server/config/prosody-overrides` first to see what the stored value was. Error: {err}"),
        }
    }

    ProsodyConfig(config)
}

// ===== Model mappings =====

pub trait IntoProsody<T> {
    fn into_prosody(self) -> T;
}

impl IntoProsody<DateLike> for db::DateLike {
    fn into_prosody(self) -> DateLike {
        match self {
            Self::Days(n) => DateLike::Days(n),
            Self::Weeks(n) => DateLike::Weeks(n),
            Self::Months(n) => DateLike::Months(n),
            Self::Years(n) => DateLike::Years(n),
        }
    }
}

impl IntoProsody<TimeLike> for db::TimeLike {
    fn into_prosody(self) -> TimeLike {
        match self {
            Self::Seconds(n) => TimeLike::Seconds(n),
            Self::Minutes(n) => TimeLike::Minutes(n),
            Self::Hours(n) => TimeLike::Hours(n),
        }
    }
}

impl<A, B> IntoProsody<Duration<B>> for db::Duration<A>
where
    A: db::DurationContent + IntoProsody<B>,
    B: DurationContent,
{
    fn into_prosody(self) -> Duration<B> {
        Duration(self.0.into_prosody())
    }
}

impl<A, B> IntoProsody<PossiblyInfinite<B>> for db::PossiblyInfinite<A>
where
    A: IntoProsody<B>,
{
    fn into_prosody(self) -> PossiblyInfinite<B> {
        match self {
            Self::Infinite => PossiblyInfinite::Infinite,
            Self::Finite(d) => PossiblyInfinite::Finite(d.into_prosody()),
        }
    }
}

impl IntoProsody<ConnectionType> for db::xmpp::XmppDirectionalConnectionType {
    fn into_prosody(self) -> ConnectionType {
        match self {
            Self::ClientToServer => ConnectionType::ClientToServer,
            Self::ServerToServerInbounds => ConnectionType::ServerToServerInbounds,
            Self::ServerToServerOutbounds => ConnectionType::ServerToServerOutbounds,
        }
    }
}
