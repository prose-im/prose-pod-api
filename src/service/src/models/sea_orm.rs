// prose-pod-api
//
// Copyright: 2025, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use std::{
    fmt::{Debug, Display},
    ops::Deref,
    str::FromStr,
};

// ===== AsString =====

pub struct SeaOrmAsString<T: Display + FromStr>(pub T);

impl<T: Display + FromStr> Display for SeaOrmAsString<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.0, f)
    }
}
impl<T: Display + FromStr> FromStr for SeaOrmAsString<T> {
    type Err = <T as FromStr>::Err;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        T::from_str(s).map(Self)
    }
}
impl<T: Display + FromStr + Debug> Debug for SeaOrmAsString<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Debug::fmt(&self.0, f)
    }
}
impl<T: Display + FromStr + Clone> Clone for SeaOrmAsString<T> {
    fn clone(&self) -> Self {
        Self(self.0.clone())
    }
}
impl<T: Display + FromStr + PartialEq> PartialEq for SeaOrmAsString<T> {
    fn eq(&self, other: &Self) -> bool {
        self.0.eq(&other.0)
    }
}
impl<T: Display + FromStr> From<T> for SeaOrmAsString<T> {
    fn from(value: T) -> Self {
        Self(value)
    }
}

impl<T: Display + FromStr> Deref for SeaOrmAsString<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: Display + FromStr> From<SeaOrmAsString<T>> for sea_orm::sea_query::Value {
    crate::sea_orm_string_value_impl!(SeaOrmAsString<T>);
}
impl<T: Display + FromStr> sea_orm::TryGetable for SeaOrmAsString<T>
where
    <T as FromStr>::Err: Debug,
{
    crate::sea_orm_try_get_by_string!(using: FromStr);
}
impl<T: Display + FromStr> sea_orm::sea_query::ValueType for SeaOrmAsString<T> {
    crate::sea_orm_string_value_type_impl!(SeaOrmAsString<T>, None, using: FromStr);
}
impl<T: Display + FromStr> sea_orm::sea_query::Nullable for SeaOrmAsString<T> {
    crate::sea_orm_string_nullable_impl!();
}

// ===== LinkedStringSet =====

#[derive(Clone, Debug, PartialEq, Eq)]
#[derive(serde::Serialize, serde::Deserialize)]
#[derive(sea_orm::FromJsonQueryResult)]
#[repr(transparent)]
pub struct LinkedStringSet(pub linked_hash_set::LinkedHashSet<String>);

impl std::ops::Deref for LinkedStringSet {
    type Target = linked_hash_set::LinkedHashSet<String>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl From<linked_hash_set::LinkedHashSet<String>> for LinkedStringSet {
    fn from(value: linked_hash_set::LinkedHashSet<String>) -> Self {
        Self(value)
    }
}

impl std::fmt::Display for LinkedStringSet {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{{")?;
        let len = self.len();
        for (i, s) in self.0.iter().enumerate() {
            write!(f, "{s:?}")?;
            if i < len - 1 {
                write!(f, ", ")?;
            }
        }
        write!(f, "}}")
    }
}
