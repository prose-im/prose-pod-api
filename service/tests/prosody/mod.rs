// prose-pod-api
//
// Copyright: 2024, RÃ©mi Bardon <remi@remibardon.name>
// License: Mozilla Public License v2.0 (MPL v2.0)

use ::service::prosody_config_from_db;
use cucumber::{given, then, when};
use insta::assert_snapshot;
use sea_orm::*;

use crate::TestWorld;

#[given("nothing has changed since the initialization of the workspace")]
fn given_nothing_changed(_world: &mut TestWorld) {
    // Do nothing, even though we could performs checks
}

#[given("every optional feature has been disabled")]
async fn given_everything_disabled(world: &mut TestWorld) -> Result<(), DbErr> {
    let mut model = world.server_config.clone().into_active_model();

    model.message_archive_enabled = Set(false);
    model.file_upload_allowed = Set(false);
    model.mfa_required = Set(false);
    model.federation_enabled = Set(false);

    world.server_config = model.update(&world.db).await?;
    Ok(())
}

#[when("generating a new Prosody configuration file from the database")]
fn when_generating_prosody_config(world: &mut TestWorld) {
    let prosody_config = prosody_config_from_db(world.server_config.clone());
    world.prosody_config = Some(prosody_config);
}

#[then(expr = "the file should match the snapshot named {string}")]
fn then_prosody_config_file_matches(world: &mut TestWorld, snapshot_name: String) {
    assert_snapshot!(snapshot_name, world.prosody_config().to_string())
}
