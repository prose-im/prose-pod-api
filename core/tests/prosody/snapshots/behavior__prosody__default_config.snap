---
source: service/tests/prosody/mod.rs
assertion_line: 39
expression: world.prosody_config().to_string()
---
-- Prose Pod Server
-- XMPP Server Configuration
-- /!\ This file has been automatically generated by Prose Pod API.
-- /!\ Do NOT edit this file manually or your changes will be overriden during the next reload.

-- Base server configuration
pidfile = "/var/run/prosody/prosody.pid"

authentication = "internal_hashed"
storage = "internal"

log = {
  info = "*console";
  warn = "*console";
  error = "*console";
}

-- Network interfaces/ports
interfaces = { "*" }
c2s_ports = { 5222 }
s2s_ports = { 5269 }
http_ports = { 5280 }
http_interfaces = { "*" }
https_ports = {}
https_interfaces = {}

-- Modules
modules_enabled = {
  "roster";
  "groups_internal";
  "saslauth";
  "tls";
  "dialback";
  "disco";
  "posix";
  "smacks";
  "private";
  "vcard_legacy";
  "vcard4";
  "version";
  "uptime";
  "time";
  "ping";
  "lastactivity";
  "pep";
  "blocklist";
  "limits";
  "carbons";
  "csi";
  "server_contact_info";
  "websocket";
  "s2s_bidi";
  "mam";
}

-- Path to SSL key and certificate for all server domains
ssl = {
  certificate = "/etc/prosody/certs/prose.org.local.crt";
  key = "/etc/prosody/certs/prose.org.local.key";
}

-- Disable in-band registrations (done through the Prose Pod Dashboard/API)
allow_registration = false

-- Mandate highest security levels
c2s_require_encryption = true
s2s_require_encryption = true
s2s_secure_auth = false

-- Enforce safety C2S/S2S limits
c2s_stanza_size_limit = 256 * 1024
s2s_stanza_size_limit = 512 * 1024

limits = {
  c2s = {
    rate = "50kb/s";
    burst = "2s";
  };
  s2sin = {
    rate = "250kb/s";
    burst = "4s";
  };
}

-- Allow reverse-proxying to WebSocket service over insecure local HTTP
consider_websocket_secure = true

-- Specify server administrator
contact_info = {
  admin = { "mailto:hostmaster@prose.org.local" };
}

-- MAM settings
archive_expires_after = "never"
default_archive_policy = true
max_archive_query_results = 100

-- Enable vCard legacy compatibility layer
upgrade_legacy_vcards = true

-- Define server members groups file
groups_file = "/etc/prosody/roster_groups.txt"

-- Server hosts and components
VirtualHost "prose.org.local"

VirtualHost "admin.prose.org.local"
  admins = { "prose-pod-api@admin.prose.org.local" }

  -- Modules
  modules_enabled = {
    "http_oauth2";
    "rest";
    "admin_rest";
    "init_admin";
  }

  -- HTTP settings
  http_host = "prose-pod-server"
  http_external_url = "http://prose-pod-server:5280"

  -- mod_init_admin
  init_admin_jid = "prose-pod-api@admin.prose.org.local"
  init_admin_password_env_var_name = "PROSE_API__ADMIN_PASSWORD"

Component "groups.prose.org.local" "muc"
  name = "Chatrooms"

  -- Modules
  modules_enabled = { "muc_mam" }

  -- MAM settings
  max_archive_query_results = 100

  restrict_room_creation = "local"

  -- MUC settings
  muc_log_all_rooms = true
  muc_log_by_default = true
  muc_log_expires_after = "never"

Component "upload.prose.org.local" "http_file_share"
  name = "HTTP File Upload"

  -- HTTP settings
  http_file_share_size_limit = 20 * 1024 * 1024
  http_file_share_daily_quota = 250 * 1024 * 1024
  http_file_share_expires_after = -1
  http_host = "localhost"
  http_external_url = "http://localhost:5280"
